**Домашнее задание 3**

В первой части третьего дз предлагается поработать с несколькими датасетами из соревнования [Boston Airbnb Open Data на Kaggle](https://www.kaggle.com/airbnb/boston). Данные можно найти на Kaggle, или по  [ссылке](https://drive.google.com/drive/folders/1lTAxxEwrCkd5WEQMJCJvBsFZc3tXgiyO?usp=sharing). Решите предложенные задания с использованием модуля Pandas (NumPy, разумеется, тоже разрешен).

Использование условий и циклов запрещено (списковые включения это тоже цикл!), за это будут обнуляться баллы.
Все задания равноценны (кроме нулевого, его выполнение необходимо для получения положительной оценки).
В каждой задаче постарайтесь придумать несколько решений. Если вы напишете несколько решений (или несколько способов сделать какую-то часть решения), сравните их производительность с помощью jupyter magics. Решением должен быть Jupyter notebook / ссылка на Colab

In [437]:
import pandas as pd
import numpy as np

0. Внимательно посмотрите на данные в таблицах. При попытке посмотреть первые несколько строк в таблице listings с помощью pandas отображаются не все колонки. Сделайте так, чтобы отображались все столбцы

    Для всех ли объявлений есть календарь? Для какой доли объявлений нет ни одного отзыва? Сколько объявлений не были доступны для сдачи ни разу за представленный период? 

In [438]:
reviews = pd.read_csv("/Users/alex/Desktop/2022/КНАД 222/Python для сбора и анализа/reviews.csv")
calendar = pd.read_csv("/Users/alex/Desktop/2022/КНАД 222/Python для сбора и анализа/calendar.csv")
listings = pd.read_csv("/Users/alex/Desktop/2022/КНАД 222/Python для сбора и анализа/listings.csv")

In [439]:
# Показать все колонки
pd.set_option('display.max_columns', None)
listings.head(5)

id                            listing_url       scrape_id  \
0  12147973  https://www.airbnb.com/rooms/12147973  20160906204935   
1   3075044   https://www.airbnb.com/rooms/3075044  20160906204935   
2      6976      https://www.airbnb.com/rooms/6976  20160906204935   
3   1436513   https://www.airbnb.com/rooms/1436513  20160906204935   
4   7651065   https://www.airbnb.com/rooms/7651065  20160906204935   

  last_scraped                                           name  \
0   2016-09-07                     Sunny Bungalow in the City   
1   2016-09-07              Charming room in pet friendly apt   
2   2016-09-07               Mexican Folk Art Haven in Boston   
3   2016-09-07  Spacious Sunny Bedroom Suite in Historic Home   
4   2016-09-07                            Come Home to Boston   

                                             summary  \
0  Cozy, sunny, family home.  Master bedroom high...   
1  Charming and quiet room in a second floor 1910...   
2  Come stay with a friendly, middle-aged guy in ...   
3  Come experience the comforts of home away from...   
4  My comfy, clean and relaxing home is one block...   

                                               space  \
0  The house has an open and cozy feel at the sam...   
1  Small but cozy and quite room with a full size...   
2  Come stay with a friendly, middle-aged guy in ...   
3  Most places you find in Boston are small howev...   
4  Clean, attractive, private room, one block fro...   

                                         description experiences_offered  \
0  Cozy, sunny, family home.  Master bedroom high...                none   
1  Charming and quiet room in a second floor 1910...                none   
2  Come stay with a friendly, middle-aged guy in ...                none   
3  Come experience the comforts of home away from...                none   
4  My comfy, clean and relaxing home is one block...                none   

                               neighborhood_overview  \
0  Roslindale is quiet, convenient and friendly. ...   
1  The room is in Roslindale, a diverse and prima...   
2  The LOCATION: Roslindale is a safe and diverse...   
3  Roslindale is a lovely little neighborhood loc...   
4  I love the proximity to downtown, the neighbor...   

                                               notes  \
0                                                NaN   
1  If you don't have a US cell phone, you can tex...   
2  I am in a scenic part of Boston with a couple ...   
3  Please be mindful of the property as it is old...   
4  I have one roommate who lives on the lower lev...   

                                             transit  \
0  The bus stop is 2 blocks away, and frequent. B...   
1  Plenty of safe street parking. Bus stops a few...   
2  PUBLIC TRANSPORTATION: From the house, quick p...   
3  There are buses that stop right in front of th...   
4  From Logan Airport  and South Station you have...   

                                              access  \
0  You will have access to 2 bedrooms, a living r...   
1  Apt has one more bedroom (which I use) and lar...   
2  I am living in the apartment during your stay,...   
3  The basement has a washer dryer and gym area. ...   
4  You will have access to the front and side por...   

                                         interaction  \
0                                                NaN   
1  If I am at home, I am likely working in my hom...   
2  ABOUT ME: I'm a laid-back, friendly, unmarried...   
3  We do live in the house therefore might be som...   
4  I love my city and really enjoy sharing it wit...   

                                         house_rules  \
0  Clean up and treat the home the way you'd like...   
1  Pet friendly but please confirm with me if the...   
2  I encourage you to use my kitchen, cooking and...   
3  - The bathroom and house are shared so please ...   
4  Please no smoking in the house, porch or on th...   

                                       thumb

In [440]:
# Для всех ли объявлений есть календарь?
pd.merge(listings, calendar, left_on='id', right_on='listing_id', how='left')['listing_id'].isna().sum()

0

#### Да, для всех объявлений есть календарь

In [441]:
# Для какой доли объявлений нет ни одного отзыва? (Ответ: доля таких объявлений)
pd.merge(listings, reviews, left_on='id', right_on='listing_id', how='left')['listing_id'].isna().sum() / listings.shape[0]

0.2108786610878661

In [442]:
# Сколько объявлений не были доступны для сдачи ни разу за представленный период? (Ответ: кол-во объявлений)
calendar['available'] = calendar['available'].apply(lambda x: x.replace('t', '1')).apply(lambda x: x.replace('f', '0')).astype(int)

In [443]:
tmp = pd.merge(listings, calendar, left_on='id', right_on='listing_id', how='left')
listings.shape[0] - tmp.groupby('id').agg({'available': 'max'}).sum()[0]

679

1. Как зовут владельца, у недвижимости которого за отчетный период была наибольшая разница между максимальной и минимальной ценой?

In [444]:
calendar['price'] = calendar['price'].apply(lambda x: str(x).replace('$', ''))
calendar['price'] = calendar['price'].apply(lambda x: str(x).replace(',', '')).astype(float)

In [445]:
listings['price'] = listings['price'].apply(lambda x: str(x).replace('$', ''))
listings['price'] = listings['price'].apply(lambda x: str(x).replace(',', '')).astype(float)
listings.rename(columns={"price": "listing_price"}, inplace=True)

In [446]:
tmp2 = pd.merge(listings, calendar.groupby('listing_id').agg({'price': 'max'}), left_on='id', right_on='listing_id', how='left')
tmp2.rename(columns={"price": "max_price"}, inplace=True)
tmp2 = pd.merge(tmp2, calendar.groupby('listing_id').agg({'price': 'min'}), left_on='id', right_on='listing_id', how='left')
tmp2.rename(columns={"price": "min_price"}, inplace=True)

In [447]:
tmp2['price_range'] = tmp2["max_price"] - tmp2["min_price"]
tmp2 = tmp2.groupby(['host_id', 'host_name']).agg({'price_range': 'max'})

In [448]:
tmp2['price_range'].sort_values(ascending=False).head(1)

host_id  host_name
2053557  Hilary       7092.0
Name: price_range, dtype: float64

#### Ответ Hilary

2. В таблице listings достаточно много самых разных колонок. В частности, есть host_listings_count, host_total_listings_count и calculated_host_listings_count. По названию понятно, что это "какое-то" число объявлений хозяина недвижимости, но какое?

    Внимание вопрос. Является ли какое-то из этих чисел количеством объявлений из listings для каждого хозяина недвижимости? Сколько в среднем объектов недвижимости у хозяина из числа указанных в listings? А сколько максимально?

    Приветствуются промежуточные вычисления и рассуждения по этому вопросу. Например, было бы круто проверить, что для всех записей одного и того же хозяина одинаковые значения в указанных выше трех столбцах

#### Итак, давайте попробуем просто проверить попарное равенство данных трех столбцов

In [449]:
listings['host_listings_count'].equals(listings['host_total_listings_count'])

True

In [450]:
listings['host_listings_count'].equals(listings['calculated_host_listings_count'])

False

#### Видно, что столбцы host_listings_count и host_total_listings_count полностью совпадают, а вот трений от них отличается. Давайте попробуем посчитать сколько же реально объявлений у каждого пользователя в данном датасете.

In [451]:
listings['calculated_host_listings_count'].equals(listings.groupby('host_id')['host_id'].transform('count'))

True

#### Таким образом, при подсчете объявлений выяснилось, что столбец "calculated_host_listings_count" содержит именно количество объявлений данного хозяина недвижимости в данном датасете. Предположим, что совпадающие столбцы 'host_listings_count' и 'host_total_listings_count' содержат общее количество объявлений хозяина недвижимости (и тех, которые не входят в датасет)

In [452]:
listings['compare'] = listings['calculated_host_listings_count'] <= listings['host_listings_count']
listings['compare'].value_counts() 

True     3583
False       2
Name: compare, dtype: int64

#### Как мы видим наша гипотеза не подтверждается на 2ух значениях, посмотрим на них поподробнее.

In [453]:
listings[listings['compare'] == False][['host_listings_count', 'host_total_listings_count', 'calculated_host_listings_count']]

host_listings_count  host_total_listings_count  \
306                     0                          0   
2852                    0                          0   

      calculated_host_listings_count  
306                                1  
2852                               1

#### Так как значения столбцов "host_listings_count" и 'host_total_listings_count' в данных строках равны 0, хотя объявления есть в датасете, кажется, что данные 2 строки просто содержат какие-то ошибки в данных. Какие-то более содержательные выводы сделать сложно)

3. Коррелируют ли средняя цена на недвижимость за всё время наблюдений и количество отзывов на неё? Иными словами, если средняя цена квартиры больше, то на неё оставляют больше или меньше отзывов?

#### Посчитаем сами по расчетным данным

In [465]:
listings['listing_price'].corr(listings['number_of_reviews'])

-0.09642705760760048

#### Посчитаем сами по датасету с отзывами

In [454]:
tmp3 = pd.merge(listings, calendar.groupby('listing_id').agg({'price': 'median'}), left_on='id', right_on='listing_id', how='left')
tmp3.rename(columns={"price": "median_price"}, inplace=True)
tmp3 = tmp3[["id", "median_price"]]

In [455]:
tmp3

id  median_price
0     12147973           NaN
1      3075044          65.0
2         6976          65.0
3      1436513          75.0
4      7651065          79.0
...        ...           ...
3580   8373729          69.0
3581  14844274         150.0
3582  14585486         117.0
3583  14603878          59.0
3584  14504422          65.0

[3585 rows x 2 columns]

In [456]:
tmp3 = pd.merge(tmp3, reviews.groupby('listing_id').agg({'id': 'count'}), left_on='id', right_on='listing_id', how='left')


In [457]:
tmp3['median_price'].corr(tmp3['id_y'])

-0.1301747475111894

#### Корреляция совсем незначительная (почти нет) -> количество отзывов не взаимосвязано со средней ценой за период

4. Давайте рассмотрим кейс, в котором имена хозяина недвижимости и гостя совпадают. Как зовут хозяйку, у которой было в качестве гостей больше всего тёзок? Что все её гости отмечали в качестве плюсов сдаваемой недвижимости?

In [458]:
tmp4 = pd.merge(reviews, listings, left_on='listing_id', right_on='id', how='left')

In [466]:
tmp4 = tmp4[['host_id', 'host_name', 'reviewer_name', 'review_scores_accuracy', 'review_scores_communication', 'review_scores_location', 'review_scores_value']] 

In [481]:
tmp5 = tmp4.loc[tmp4['host_name'] == tmp4['reviewer_name']].groupby('host_id').agg({'reviewer_name': 'count'})

In [484]:
tmp4 = pd.merge(tmp4, tmp5, on='host_id')

In [504]:
tmp4.groupby(['host_id', 'host_name']).agg({"reviewer_name_y": 'max', 'review_scores_accuracy': 'median', "review_scores_communication": 'median', 'review_scores_location': 'median', 'review_scores_value': 'median'}).sort_values(by="reviewer_name_y", ascending=False)

reviewer_name_y  review_scores_accuracy  \
host_id  host_name                                            
85770    Robert                   3                     9.0   
1444340  Will                     3                     9.0   
5618949  Alan                     3                     9.0   
9898255  Jessica                  3                     9.0   
2503703  Susan                    3                    10.0   
...                             ...                     ...   
11272292 Rebecca                  1                     9.0   
11631886 Cynthia                  1                     9.0   
13382900 Bernie                   1                    10.0   
13929879 David                    1                     9.0   
68072593 James                    1                     9.0   

                    review_scores_communication  review_scores_location  \
host_id  host_name                                                        
85770    Robert                             9.0                    10.0   
1444340  Will                               9.0                     9.0   
5618949  Alan                              10.0                    10.0   
9898255  Jessica                           10.0                     9.0   
2503703  Susan                             10.0                    10.0   
...                                         ...                     ...   
11272292 Rebecca                            9.0                     9.0   
11631886 Cynthia                           10.0                     8.0   
13382900 Bernie                            10.0                    10.0   
13929879 David                             10.0                     9.0   
68072593 James                              9.0                    10.0   

                    review_scores_value  
host_id  host_name                       
85770    Robert                     9.0  
1444340  Will                       8.0  
5618949  Alan                       9.0  
9898255  Jessica                    9.0  
2503703  Susan                      9.0  
...                                 ...  
11272292 Rebecca                    9.0  
11631886 Cynthia                    9.0  
13382900 Bernie                    10.0  
13929879 David                      9.0  
68072593 James                      8.0  

[77 rows x 5 columns]

#### Как мы видим, есть сразу несколько хозяинов, у которых было по 3ое (максимальное кол-во) тезок в качестве гостей. Выберем произвольного под id 85770 с именем Robert. Так как из таблицы reviews извлечь ничего дельного относительно плюсов недвижимости нельзя, то вернемся к таблице с данными только по объявлениям. Из таблицы выше можно видеть, что ВСЕ гости отмечают отличную локацию сдаваемой недвижимости, а также ставили высокие баллы и по другим критериям.